### ISSUE TITLE
Continue stripping psyGen nodes

### ISSUE THREAD

**Issue by sergisiso** [stfc/PSyclone#1208]
When I start pulling Node concepts out of psyGen I didn't finish it because it was not clear what the boundary was between core nodes and psyclone domain concepts (at that time the separation was between Kernel nodes and Invoke nodes).

I think now we have a clearer structure between language-level nodes, and high-level concept nodes. Language level nodes are concepts which exist on C/Fortran and should be placed in `psyclone.psyir.nodes`, each backend in `psyclone.psyir.backends` should have a visitor for it and they don't need lowering to other PSyIR concepts. This should make the frontend->nodes->backend more or less a consistent set of 1-to-1 mappings.

Higher level "PSyclone" concepts should be in `psyclone.domain.common` and they must implement `lower_to_language_level()` method if their base node visitor doesn't work out of the box.

Therefore:
- psyGen.py: `Directive`, `ACC*Directive` and `OMP*Directive` should be moved to `psyclone.psyir.nodes`
- psyGen.py: InvokeSchedule, GlobalSum, HaloExchange, Kern, CodedKern, InlinedKern, BuiltIn should be moved to `psyclone.domain.common.[psylayer/nodes]`
- `psyclone.psyir.nodes` PSyDataNode Family (nan_test_node, profile_node, ...) should also be moved to `psyclone.domain.common.[psylayer/nodes]`

I am right in thinking all of these are exclusive to the PSy-layer?

Note that still the API specialisation of all of them wherever they come from will be group together at the same place (e.g. GOLoop or GOGlobalSum will be under `psyclone.domain.gocean.?` )

@rupertford @arporter @hiker @TeranIvy does this looks sensible to all of you?

**Comment by arporter** [comment]
Yes, that looks sensible to me.

**Comment by arporter** [comment]
Just a note to say that it looks as though `Kern` doesn't have a method to get its body (cf `Loop.loop_body`, `Directive.dir_body`). When you come to move `Kern` it may be worth adding this?

**Comment by sergisiso** [comment]
If I understand correctly, shouldn't this be only in InlinedKern as CodedKern and BulitIn don't have a body?

**Comment by arporter** [comment]
Ah yes, that's true, although we do have something like 'get_kernel_schedule' that effectively does the same thing for CodedKern. Given that that's reading from a separate file though it may be best to keep that as is? In fact, BuiltIn could (possibly should) have a body. Once we move to generating the PSy layer using PSyIR then we will have to have PSyIR for each Builtin.

**Comment by rupertford** [comment]
A couple of comments. 1) We might want the common version to have an abstract `lower_to_language_level()`. I'm thinking about the halo exchanges here, where the implementation will be specific to the API, but the concept is common. 2) I don't think all the common PSyIR nodes should go in `psylayer/nodes`. It seems to me that the halo exchange nodes (and global sum) are more generic concepts, outside of the algorithm/psy/kern structure.

**Comment by sergisiso** [comment]
>  1) We might want the common version to have an abstract lower_to_language_level().

That's fine, common PSyclone nodes can be abstract. The restriction is that they don't appear directly on the backend because they are higher level than what a language provides.

> 2) I don't think all the common PSyIR nodes should go in psylayer/nodes. It seems to me that the halo exchange nodes (and global sum) are more generic concepts, outside of the algorithm/psy/kern structure.

This may be worth discussing in a call, to me they fit better as PSyclone psy-layer nodes given that they are not language concepts and for how they are used (they have fields and args as methods...)

**Comment by rupertford** [comment]
> This may be worth discussing in a call, to me they fit better as PSyclone psy-layer nodes given that they are not language concepts and for how they are used (they have fields and args as methods...)

Very happy to discuss. I take your point about their use in the PSy layer but I still see them as PSyclone/common concepts. If someone wanted to use PSyclone to just modify halo exchanges in a code they might not want and should not need to be concerned about algorithm/psy/kernel layers.

**Comment by rupertford** [comment]
After discussion, we actually agree. The "/" in "psylayer/nodes" was an or, not a sub-directory.

**Comment by arporter** [comment]
I noticed today that we still have `psyGen::InlinedKern` but it is now only ever used in tests. It should be straightforward to remove it.

**Comment by sergisiso** [comment]
Maybe we need to discuss how we apply InlineTrans to kernels first?

The lowering during the script followed by the InlineTrans is not a very nice solution. You can not do other transformations involing kernels after that. Maybe the concept of InlinedKern would still make sense?

**Comment by arporter** [comment]
Yes, a discussion would be good: not having to do lowering in the script would be much nicer.
